modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java

```diff
@@ -226,26 +233,59 @@ public CompletableFuture<QueryPlan> prepareAsync(
             return CompletableFuture.failedFuture(new SchemaNotFoundException(schemaName));
         }
 
+        // Add an action to trigger planner timeout, when operation times out.
+        // Or trigger timeout immediately if operation has already timed out.
+        QueryCancel cancelHandler = operationContext.cancel();
+        assert cancelHandler != null;
+
+        long actualPlannerTimeout;
+        Instant deadline = cancelHandler.deadline();
+        if (deadline == null) {
+            actualPlannerTimeout = plannerTimeout;
+        } else {
+            Instant now = Instant.now();
+            long timeout = Duration.between(now, deadline).toMillis();
+
+            actualPlannerTimeout = Math.min(timeout, plannerTimeout);
```

Comment:
I have doubts about this.
We have different timeouts
1. query timeout
2. planner timeout

And at this point we already scheduled query timeout. When query timeout will happen we will abort planner and we will raise error about query timeout.

Why do we also need to reduce the planner timeout?
It looks like we may be throwing the wrong error, that a  planner timeout occurred, but the user needs to increase the request timeout.

Reply:
Fixed.